helm-v3/deisgn-proposal.md

```diff
@@ -0,0 +1,715 @@
+# Helm 3 Design Proposal
+
+This document contains a proposal for Helm 3. It assumes deep familiarity with Helm 2, and assumes that much of Helm 2's architecture will simply carry over without major change.
+
+## Summary
+
+- Tiller is gone, and there is only one functional component (`helm`)
+- Charts are updated with a new Chart.yaml, overlays, libraries, schematized values, and the ext directory
+- Helm will use a "lifecycle events" emitter/handler model.
+- Helm has an embedded Lua engine for scripting some event handlers. Scripts are stored in charts.
+- State is maintained with three CRDs: Application CRD, Release, and ReleaseVersion. Neither Helm CRD has a controller.
+- Resources created by hooks will now be managed
+- For pull-based DevOps workflow, a new Helm Controller project will be started
+- Cross platform plugins in Lua that only have a runtime dependency on Helm
+- A complementary command to `helm fetch` to push packages to a repository
+
+## The Basic Architecture
+
+Helm 3 is a single-service architecture. One executable is responsible for implementing Helm. There is no client/server split, nor is the core processing logic distributed among components.
+
+The reference implementation of Helm 3 is a single command-line client with no in-cluster server or controller. This tool exposes command-line operations, and unilaterally handles the package management process.
+
+The reference implementation has two distinct parts:
+
+1. The command line façade, which translates commands, subcommands, flags, and arguments into a Helm operation
+2. The Helm library, which provides the logic for executing all Helm operations.
+
+By design, the Helm library _must_ be usable as a standalone library.
+
+In Appendix A, we specify a second implementation of Helm 3 that provides a controller-oriented model with no client. The Helm Controller provides a CRD-based façade to receive commands, but executes them via the Helm library.
```

Comment:
No commitments yet on timelines. Basically, it will come down to who is available for doing what work. That approach has some additional complexities that must be worked out, so at the moment it looks like CLI first, controller second.